File: entity-meta-generator-lib/src/main/java/com/loudsight/meta/serialization/EntityType.java
package com.loudsight.meta.serialization;

import java.util.HashMap;
import java.util.Map;

/**
 * Enum representing different entity types for serialization.
 */
public enum EntityType {
    /**
     * Null type.
     */
    NULL('a'),
    /**
     * Byte type.
     */
    BYTE('b'),
    /**
     * Bytes type.
     */
    BYTES('c'),
    /**
     * Integer type.
     */
    INTEGER('d'),
    /**
     * Long type.
     */
    LONG('e'),
    /**
     * String type.
     */
    STRING('f'),
    /**
     * Double type.
     */
    DOUBLE('g'),
    /**
     * Float type.
     */
    FLOAT('h'),
    /**
     * List type.
     */
    LIST('i'),
    /**
     * Set type.
     */
    SET('j'),
    /**
     * Array type.
     */
    ARRAY('k'),
    /**
     * Map type.
     */
    MAP('l'),
    /**
     * Boolean type.
     */
    BOOLEAN('m'),
    /**
     * Enum type.
     */
    ENUM('n'),
    /**
     * Custom type.
     */
    CUSTOM('o'),
    /**
     * Class type.
     */
    CLASS('p'),
    /**
     * DateTime type.
     */
    DATETIME('q');
    /**
     * The byte code for this entity type.
     */
    private final byte code;

    /**
     * Constructs an EntityType with the given character code.
     * @param code the character code
     */
    EntityType(char code) {
        this.code = (byte)code;
    }

    private static final Map<Byte, EntityType> codeToEntityType = new HashMap<>();

    static {

        for (var entityType : values()) {
            codeToEntityType.put(entityType.code, entityType);
        }
    }

    /**
     * Gets the byte code for this entity type.
     * @return the byte code
     */
    public byte getCode() {
        return code;
    }

    /**
     * Gets the EntityType for the given byte code.
     * @param code the byte code
     * @return the EntityType
     */
    public static EntityType getEntityType(byte code) {
        var res = codeToEntityType.get(code);

        if (res == null) {
            throw new IllegalArgumentException("Unknown entity type code: " + code);
        }

        return res;
    }
}

File: entity-meta-generator-lib/src/main/kotlin/com/loudsight/meta/serialization/EntityType.kt
package com.loudsight.meta.serialization

enum class EntityType(val code: Byte) {
    NULL('a'),
    BYTE('b'),
    BYTES('c'),
    INTEGER('d'),
    LONG('e'),
    STRING('f'),
    DOUBLE('g'),
    FLOAT('h'),
    LIST('i'),
    SET('j'),
    ARRAY('k'),
    MAP('l'),
    BOOLEAN('m'),
    ENUM('n'),
    CUSTOM('o'),
    CLASS('p'),
    DATETIME('q')
    ;

    constructor(code: Char) : this(code.code.toByte())

    companion object {
        val codeToEntityType = HashMap<Char, EntityType>()

        init {
            for (entityType in values()) {
                codeToEntityType[entityType.code.toChar()] = entityType
            }
        }
        fun getEntityType(code: Char): EntityType {
            return codeToEntityType[code] ?: throw IllegalArgumentException("Unknown entity type code: $code")
        }
    }
}